File: service/image_batch_test.go
package service

import (
	"context"
	"strings"
	"testing"
	"time"

	"github.com/thaodangspace/money-bot/ai"
	"github.com/thaodangspace/money-bot/domain"
)

type batchTestAI struct{}

func (batchTestAI) ParseTransaction(context.Context, string) (domain.Transaction, error) {
	return domain.Transaction{}, nil
}

func (batchTestAI) ParseImageTransactions(context.Context, string, string, []byte) (ai.ImageTransactionExtraction, error) {
	return ai.ImageTransactionExtraction{
		Kind:     ai.ImageExtractionList,
		Detected: 3,
		Transactions: []domain.Transaction{
			{Type: domain.TransactionExpense, Category: "drink", Note: "Cafe", Amount: 50000},
			{Type: domain.TransactionExpense, Category: "transport", Note: "Fuel", Amount: 120000},
			{Type: domain.TransactionExpense, Category: "food", Note: "Lunch", Amount: 75000},
		},
	}, nil
}

type batchTestLedger struct {
	transactions []domain.Transaction
	updates      []int
}

func (l *batchTestLedger) AppendTransactions(_ context.Context, updateID int, transactions []domain.Transaction) (AppendBatchResult, error) {
	l.updates = append(l.updates, updateID)
	l.transactions = append(l.transactions, transactions...)
	return AppendBatchResult{Status: AppendWritten}, nil
}

func (l *batchTestLedger) MonthlySummary(context.Context, int, time.Month) (domain.MonthlySummary, error) {
	return domain.MonthlySummary{}, nil
}

func TestImageBatchPreviewAndConfirmationWritesAllTransactions(t *testing.T) {
	ledger := &batchTestLedger{}
	svc := mustService(t, ledger, batchTestAI{}, nil, fixedClock())
	prepared, err := svc.PrepareImage(context.Background(), 77, ImageInput{MIMEType: "image/png", Data: []byte{1}})
	if err != nil || prepared.Token == "" {
		t.Fatalf("prepared=%#v err=%v", prepared, err)
	}
	if len(ledger.transactions) != 0 || !strings.Contains(prepared.Text, "3 giao dịch") || !strings.Contains(prepared.Text, "1. Chi tiêu") || !strings.Contains(prepared.Text, "Tổng chi tiêu: 245.000") {
		t.Fatalf("preview=%q writes=%#v", prepared.Text, ledger.transactions)
	}
	result, err := svc.ConfirmImage(context.Background(), prepared.Token)
	if err != nil || !result.Parsed || len(ledger.transactions) != 3 || len(ledger.updates) != 1 {
		t.Fatalf("result=%#v err=%v ledger=%#v", result, err, ledger)
	}
	for _, tx := range ledger.transactions {
		if tx.SourceUpdateID != 77 || tx.Date.IsZero() {
			t.Fatalf("transaction=%#v", tx)
		}
	}
}

File: src/service/image_batch_test.ts
import type { ImageTransactionExtraction } from '../adapters/ai/image_types.ts';
import { MoneyService } from './money_service.ts';
import type { AIParser, AppendBatchResult, Ledger } from './types.ts';
import type { MonthlySummary } from '../domain/summary.ts';
import type { Transaction } from '../domain/transaction.ts';

class BatchAI implements AIParser {
  parseTransaction(): Promise<Transaction> {
    return Promise.reject(new Error('unused'));
  }
  parseImageTransactions(): Promise<ImageTransactionExtraction> {
    return Promise.resolve({
      kind: 'transaction_list',
      detected: 3,
      transactions: [
        { type: 'expense', category: 'drink', amount: 50_000, note: 'Cafe', date: '2026-07-18' },
        {
          type: 'expense',
          category: 'transport',
          amount: 120_000,
          note: 'Fuel',
          date: '2026-07-18',
        },
        { type: 'expense', category: 'food', amount: 75_000, note: 'Lunch', date: '2026-07-18' },
      ],
    });
  }
}

class BatchLedger implements Ledger {
  appended: Transaction[] = [];
  updates: number[] = [];
  appendTransactions(
    _signal: AbortSignal,
    updateId: number,
    transactions: Transaction[],
  ): Promise<AppendBatchResult> {
    this.updates.push(updateId);
    this.appended.push(...transactions);
    return Promise.resolve({ status: 'written', targetSheets: ['2026-07'] });
  }
  monthlySummary(): Promise<MonthlySummary> {
    return Promise.resolve({
      year: 2026,
      month: 7,
      totalExpenses: 0,
      totalIncome: 0,
      balance: 0,
      entryCount: 0,
    });
  }
}

Deno.test('image list previews and confirms all transactions in one write', async () => {
  const ledger = new BatchLedger();
  const service = new MoneyService({
    ledger,
    ai: new BatchAI(),
    clock: () => new Date('2026-07-18T10:00:00Z'),
  });
  const prepared = await service.prepareImage(new AbortController().signal, 77, {
    caption: '',
    mimeType: 'image/png',
    data: new Uint8Array([1]),
  });
  if (
    !prepared.text.includes('3 giao dịch') || !prepared.text.includes('18/07/2026') ||
    !prepared.text.includes('Tổng chi tiêu: 245.000')
  ) throw new Error(prepared.text);
  const previewWrites = ledger.appended.length;
  if (previewWrites !== 0) throw new Error('preview wrote to ledger');
  const result = await service.confirmImage(new AbortController().signal, prepared.token);
  const writeCount = ledger.appended.length;
  const updateCount = ledger.updates.length;
  if (!result.parsed || writeCount !== 3 || updateCount !== 1) {
    throw new Error(JSON.stringify({ result, ledger }));
  }
  if (ledger.appended.some((transaction) => transaction.sourceUpdateId !== 77)) {
    throw new Error('source update ID was not preserved');
  }
});
